# **February Phase 2 OIB-SIP Internship-Data Science**
*By-Karan Sardar*

***Problem Statement:***
*Weve all been the recipient of spam emails before. Spam mail, or junk mail, is a type of email
that is sent to a massive number of users at one time, frequently containing cryptic
messages, scams, or most dangerously, phishing content.*

*In this Project, use Python to build an email spam detector. Then, use machine learning to
train the spam detector to recognize and classify emails into spam and non-spam. Lets get
started.*

# **EMAIL SPAM CLASSIFICATION**
#TASK 4

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import chardet
import warnings
warnings.filterwarnings('ignore')

In [2]:
with open('spam.csv', 'rb') as f:
    result = chardet.detect(f.read())
df = pd.read_csv('spam.csv', encoding=result['encoding'])
df.head()


v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [3]:
df.drop(['Unnamed: 2','Unnamed: 3','Unnamed: 4'],axis=1,inplace=True)

In [4]:
df.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [5]:
df.groupby('v1').describe()

v2                                                               
     count unique                                                top freq
v1                                                                       
ham   4825   4516                             Sorry, I'll call later   30
spam   747    653  Please call our customer service representativ...    4

In [6]:
df.dtypes

v1    object
v2    object
dtype: object

In [7]:
df["v1"] = df["v1"].astype('category')
df.dtypes

v1    category
v2      object
dtype: object

In [8]:
df["v1"] = df["v1"].cat.codes
df.head()

v1                                                 v2
0   0  Go until jurong point, crazy.. Available only ...
1   0                      Ok lar... Joking wif u oni...
2   1  Free entry in 2 a wkly comp to win FA Cup fina...
3   0  U dun say so early hor... U c already then say...
4   0  Nah I don't think he goes to usf, he lives aro...

In [9]:
doc = df.iloc[:,1]
doc
#doc2=doc.to_numpy().reshape(-1,1)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                Will Ì_ b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: v2, Length: 5572, dtype: object

In [10]:
#Count Vectorizing : Extracting text
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(doc)
vectorizer.get_feature_names_out()
vector = vectorizer.transform(doc)
bet=vector.toarray()

In [11]:
#train test split to divide training and testing data
v1=df.iloc[:,0]
v2=v1.to_numpy()
v3=v2.reshape(-1,1)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(bet, v3, train_size = 0.7, test_size = 0.3, random_state = 43 )

# **LOGISTIC REGRESSION**

In [12]:
from sklearn.linear_model import LogisticRegression
logi=LogisticRegression()

In [13]:
logi.fit(X_train,y_train)

LogisticRegression()

In [14]:
y_pred=logi.predict(X_test)
print(y_pred)

[0 0 0 ... 0 0 0]


In [15]:
score=logi.score(X_test,y_test)
score2=logi.score(X_train,y_train)
print('Accuracy score on train is:',score2*100)
print('Accuracy score on test is:',score*100)

Accuracy score on train is: 99.74358974358975
Accuracy score on test is: 97.78708133971293


In [16]:
from sklearn.metrics import classification_report, confusion_matrix
confu=confusion_matrix(y_test,y_pred)
print(confu)

[[1434    6]
 [  31  201]]


In [17]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99      1440
           1       0.97      0.87      0.92       232

    accuracy                           0.98      1672
   macro avg       0.97      0.93      0.95      1672
weighted avg       0.98      0.98      0.98      1672



# **DECISION TREE**

In [18]:
#importing decision tree model
from sklearn.tree import DecisionTreeClassifier
dec=DecisionTreeClassifier(criterion='gini')
dec.fit(X_train,y_train)

DecisionTreeClassifier()

In [19]:
y_pred2=dec.predict(X_test)
y_pred2

array([0, 0, 0, ..., 0, 0, 0], dtype=int8)

In [20]:
print(classification_report(y_test, y_pred2))

              precision    recall  f1-score   support

           0       0.97      0.98      0.98      1440
           1       0.88      0.82      0.85       232

    accuracy                           0.96      1672
   macro avg       0.93      0.90      0.91      1672
weighted avg       0.96      0.96      0.96      1672



In [21]:
confu2=confusion_matrix(y_test,y_pred2)
print(confu2)

[[1415   25]
 [  41  191]]


# **RANDOM FOREST**

In [22]:
#importing Random forest model
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier(n_estimators=100)
rf.fit(X_train,y_train)

RandomForestClassifier()

In [23]:
y_pred3=rf.predict(X_test)
y_pred3

array([0, 0, 0, ..., 0, 0, 0], dtype=int8)

In [24]:
print(classification_report(y_test, y_pred3))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98      1440
           1       0.99      0.77      0.87       232

    accuracy                           0.97      1672
   macro avg       0.98      0.89      0.92      1672
weighted avg       0.97      0.97      0.97      1672



In [25]:
confu3=confusion_matrix(y_test,y_pred3)
print(confu3)

[[1438    2]
 [  53  179]]


# **LIGHT GBM**

In [26]:
import lightgbm as lgb
light=lgb.LGBMClassifier()

In [27]:
light.fit(X_train,y_train)

[LightGBM] [Info] Number of positive: 515, number of negative: 3385
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.019564 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1213
[LightGBM] [Info] Number of data points in the train set: 3900, number of used features: 410
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.132051 -> initscore=-1.882942
[LightGBM] [Info] Start training from score -1.882942


LGBMClassifier()

In [28]:
y_pred4=light.predict(X_test)
y_pred4

array([0, 0, 0, ..., 0, 0, 0], dtype=int8)

In [29]:
confu4=confusion_matrix(y_test,y_pred4)
print(confu4)

[[1437    3]
 [  29  203]]


In [30]:
print(classification_report(y_test, y_pred4))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99      1440
           1       0.99      0.88      0.93       232

    accuracy                           0.98      1672
   macro avg       0.98      0.94      0.96      1672
weighted avg       0.98      0.98      0.98      1672



# **LIGHTGBM HYPERPARAMETER TUNING**

In [33]:
params = {
"num_leaves": [31, 63, 127],
"max_depth": [-1, 3, 5],
"colsample_bytree": [0.8, 1.0],
"verbose":[-1]
}

In [34]:
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(light, param_grid=params, cv=5, scoring="accuracy")
grid_search.fit(X_train,y_train)

GridSearchCV(cv=5, estimator=LGBMClassifier(),
             param_grid={'colsample_bytree': [0.8, 1.0],
                         'max_depth': [-1, 3, 5], 'num_leaves': [31, 63, 127],
                         'verbose': [-1]},
             scoring='accuracy')

In [35]:
best_parameter=grid_search.best_params_
best_parameter

{'colsample_bytree': 0.8, 'max_depth': -1, 'num_leaves': 31, 'verbose': -1}

In [36]:
lgbm=lgb.LGBMClassifier(colsample_bytree= 0.8, max_depth= -1, num_leaves=31)
lgbm.fit(X_train,y_train)

[LightGBM] [Info] Number of positive: 515, number of negative: 3385
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.013999 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1213
[LightGBM] [Info] Number of data points in the train set: 3900, number of used features: 410
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.132051 -> initscore=-1.882942
[LightGBM] [Info] Start training from score -1.882942


LGBMClassifier(colsample_bytree=0.8)

In [37]:
y_pred5=lgbm.predict(X_test)
y_pred4

array([0, 0, 0, ..., 0, 0, 0], dtype=int8)

In [38]:
confu_lgbm=confusion_matrix(y_test,y_pred4)
print(confu_lgbm)

[[1437    3]
 [  29  203]]


In [39]:
print(classification_report(y_test, y_pred5))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99      1440
           1       0.99      0.89      0.94       232

    accuracy                           0.98      1672
   macro avg       0.98      0.95      0.96      1672
weighted avg       0.98      0.98      0.98      1672



**Hyperparameter tuning the LightGBM model, gave a 98% accuracy on test data. Thank you!!**